### ISSUE TITLE
Add ES6 module support

### ISSUE THREAD

**Issue by jplaisted** [google/closure-library#846]
We've been discussing ES6 module support. [Here](https://docs.google.com/document/d/1Qked4pyfj5qYbsJspiBKfvzh5t3HfuhN92AdCPTAR1I/edit?usp=sharing) is the current design for library and compiler work.

tl;dr

- goog.require will support relative paths **for goog and ES6 modules from goog modules and ES6 modules** (though ES6 -> ES6 should use import not goog.require).
- ES6 modules will need to be transpiled to goog modules when whitespace bundling. Currently whitespace bundling is not supported for ES6 modules.
- The above two bullets combine so that when whitespace bundling mixed Closure / ES6 module code only ES6 modules need to be _lightly_ transpiled to goog modules. Imported paths will become goog.require'd paths.
- There will be a new goog.bootstrap function as a replacement for "bootstrap goog.requires" (top level gog.requires when loading code via the debug loader). Because loading will be asynchronous there needs to be an alternative to goog.require that provides a callback when things have loaded.

**Comment by jplaisted** [comment]
Compiler tracking issue is here. https://github.com/google/closure-compiler/issues/2658

**Comment by bowbahdoe** [comment]
Its unclear to me if the new ES6 module support will be usable from outside the closure compiler. As of right now there is a way to use the closure library from node.js, but that relies on a bootstrap script which uses the 'fs' module, and so is not suited for regular bundling with alternative tools such as browserify and webpack. Is there a global polyfil that could make goog.require and good.provide parse correctly, or is there a way to "convert" the closure library on npm to use regular es6/commonjs modules?

**Comment by jplaisted** [comment]
I'm not sure what you're asking here. Are you worried about the Closure library moving to ES6 modules and how that will work with node.js? This bug isn't about moving Closure library to ES6 modules, just adding support for people that wish to use ES6 modules in their own library that depends Closure. The former is further off than the latter due to compatibility issues, both internally and with node.

So yes if someone writes a Javascript library using ES6 modules and Closure it won't be able to be loaded by node without some kind of transformation. But that's not really on us.



Also per some discussion internally and in the compiler issue we're thinking of transpiling ES6 modules to some $jscomp.module instead, which will be pollyfiled. They'll act similarly to goog.module but without the dependency on the Closure library. Of course in a full compile these won't exist (they'll be transformed again), only in bundles and per file transpilation.

So the above statement:

> The above two bullets combine so that when whitespace bundling mixed Closure / ES6 module code only ES6 modules need to be lightly transpiled to goog modules. Imported paths will become goog.require'd paths.

Is incorrect and should now be:

> The above two bullets combine so that when whitespace bundling ES6 module code ES6 modules need to be lightly transpiled to $jscomp.moduels. Imported paths will become $jscomp.import'd paths.

Apologies for not updating this issue, but I did update the compiler issue and the document. Though I think this is unrelated to your question.

**Comment by bowbahdoe** [comment]
You answered my question more or less. I was wondering how a person in the node ecosystem could use the closure library. Right now the package on npm is useable from node.js, but not from code compiled from the frontend using that ecosystem. I was wondering if this issue implied that the current system on nodejs for polyfilling goog.require and goog.provide would be updated so that it could be used on the frontend without the Google closure compiler. Based on your response, I gather that transpiling the closure library to commonjs (eliminating the need for the fs bootstrap) is something that is on the radar.

**Comment by jplaisted** [comment]
What issues are you getting? Closure Library doesn't use ES6 modules yet, I assume it is your code?

base.js should support ES6 modules today (as of d786e236e303707ddd49b697480935cf274100ee), along with the compiler. However some of the tools aren't update (like closurebuilder.py and depswriter.py).

**Comment by jplaisted** [comment]
I'd ask @ChadKillingsworth about how to use Closure Library + ES6 modules + Browserify, I think he's done a bit of work with that. Though I could be wrong.

**Comment by BusbyActual** [comment]
@jplaisted perhaps I'm mistaken about the library being the problem. I'll check in with Chad if he's available. This issue is goog is not defined after I require it.  This library is dependent on closure and doesn't specify how to use it.

Example:

**app.js**
```
require('base.js');
require('blockly_compressed');
require('blockly/logic.js');
```

**Uncaught ReferenceError: goog is not defined** 
```
// references logic.js these lines
goog.provide('Blockly.Blocks.logic');
goog.require('Blockly.Blocks');
```

**Comment by jplaisted** [comment]
blockly_compressed contains Closure's base.js already if you look at it, it defines `goog`. It contains some of blockly (not sure what pieces) and its dependencies, including parts of Closure. I don't know why `goog` isn't defined but I'm not super familiar with Browserify.

I'd also maybe ping the blockly folks and see what their recommended workflow is. I think in your instance since you want to use Closure too you shouldn't include their precompiled / compressed files. You should include their raw sources you need. That way Closure itself is only loaded once.

I don't think the work done for and todo with this issue will help you at all if you're using Browserify.


Speaking of which reading @bowbahdoe 's older comments now make more sense to me now. I was perhaps mislead in my "CommonJS like" statements. I was just discussing the underlying implementation details, and they aren't actually CommonJS modules - just basded on them. And this issue has nothing to do with Closure Library itself transforming to CommonJS or anything like that. This bug is / was all about changing the code loader in `base.js` to load ES6 modules, the transpiler in `transpile.js` to be able to handle ES6 modules, and for some interop via `goog.require`. And the design does require loading through `base.js` or the Closure Compiler and sadly is not (currently) compatible with Browserify or Webpack or the like. You'll have to go through our workflows for now, this issue isn't about adding that support, though it'd be nice to.

This bug is **not** about moving Closure Library to ES6 modules, which would be nice. Adding the support described above is a prerequisite to an incremental migration. There's a lot of other blockers, however, and I personally think a migration very far off, if ever.

**tl;dr** this issue is about adding ES6 support to the Closure Debug Loader and Compiler, sorry for any confusion around that. It is not about adding support outside of that, e.g. to Browserify.

**Comment by BusbyActual** [comment]
@jplaisted  I appreciate you for your insight into the issue. You've been a great help cheers.

**Comment by jplaisted** [comment]
It isn't possible to wrap ES6 modules. `import` and `export` statements must be at the top level of a module. We transform them to a CommonJS-like module if we need to (e.g. for a bundle or transpilation). Closure's code loader in closure/goog/base.js will attempt to load them untransformed, if possible.

I'll admit I don't know much about Karma. Does it create a bundle, does it try to load files "raw" (in a `<script src="filename.js></script>` tag), or does it let the Closure code loader in closure/goog/base.js load the code (when using the karma-closure plugin)? Depending on the answer to this how to handle ES6 module <=> Closure interop will be different.

Also fwiw Karma itself does not support ES6 modules. See https://github.com/karma-runner/karma/issues/2903. I'm guessing this will complicate things.


To answer this specific question:

> Looking at the design doc, I'm not exactly sure of the recommended way to do runtime transpilation for ES6 modules, directly in the browser, without any bundling phase

If you can go through Closure's code loader it will do the runtime transpilation for you, if needed. "If needed" currently means that `'noModule' in document.createElement('script')` returns `false`.

**Comment by jplaisted** [comment]
>Simply put, once I get the contents of an ES6 module as a string, is there any API provided by base.js that can parse the raw JS and do the Closure→ES6 transpilation?

Not publicly, no. The idea is that you should go through Closure's debug code loader or provide a bundle, preferably the latter (the debug code loader is pretty old and non-standard at this point!). If you are using Closure's code loader it can transform ES6 modules on browsers that do not support them.

If you transform your ES6 modules, either by bundling or serving on a per file basis before serving, then Closure will need to be made aware of them. Closure needs to know about them while they are executing so that `goog.module.declareNamespace` works correctly and so that `goog.require` will return a value inside the module.

If you take a look at the implementation of `goog.module.declareNamespace` you can see how base expects this to be done. It expects a module runtime on the global `$jscomp` object that acts like CommonJS so it can get a reference to the exports.

```js
   // ...
   // Bundled - not debug loading, no module loader state.
    var jscomp = goog.global['$jscomp'];
    if (!jscomp || typeof jscomp.getCurrentModulePath != 'function') {
      throw new Error(
          'Module with namespace "' + namespace +
          '" has been loaded incorrectly.');
    }
    var exports = jscomp.require(jscomp.getCurrentModulePath());
    // ...
```


If you want to load the ES6 modules raw (e.g. `<script type="module">`) then things are a little more complicated. The current design assumes you are going through the `goog.Dependency` interface which allows you to register ES6 module exports via `goog.LoadController.prototype.registerEs6ModuleExports`. But if you're loading code through Karma you won't be going through this interface. I think its still possible to use this interface, though a little messy.

So if we wanted to load this file with Karma:

```js
// examplemodule.js
goog.module.declareNamespace('example.module');
const array = goog.require('goog.array');
export const filtered = array.filter([0, 1, 2, 3], (e) => e  % 2);
```

You could write a bootstrap file like so:
```js
// bootstrap.js
// Singleton dependency that is used to register ES6 module exports and nothing more.
const karmaEs6Dependency = new class extends goog.Dependency {
  constructor() {
    super('', '', [], [], {});
  }
  preLoadingModule(path) {
    this.path = path;
    this.loadingModule = true;
    // Fake a call to goog.addDependency, which in turn will make "path"
    // require-able. When require'd it will call the createDependency method
    // of the DepenencyFactory, which we've set to just return this object. Then
    // it will call this object's load method.
    goog.addDependency(path, [path], []);
    goog.require(path);
    this.controller.setModuleState(goog.ModuleType.ES6);
  }
  registerModule(exports, namespace = undefined) {
    // namespace is the argument of goog.module.declareNamespace. Optional, only if called
    // in the module.
    this.controller.registerEs6ModuleExports(this.path, exports, namespace);
    this.loadingModule = false;
  }
  postLoadingModule() {
    this.controller.clearModuleState();
    this.controller.loaded();
  }
  load(controller) {
    if (!this.loadingModule) throw new Error();
    this.controller = controller;
  }
};

class KarmaDependencyFactory extends goog.DependencyFactory {
  createDependency() { return karmaEs6Dependency; }
}

goog.setDependencyFactory(new KarmaDependencyFactory());
```

And then ensure your html generator writes something like:
```html
<script src="goog/base.js"></script>
<script src="bootstrap.js"></script>
<!-- other closure files (dependencies of goog.array) !-->
<script src="goog/array/array.js"></script>
<!-- 
  Use 3 different script tags so that if the middle module errors
  (which can happen if any dependencies error) then the state
  is still cleaned up.
-->
<script type="module">
  karmaEs6Dependency.preLoadModule('/examplemodule.js');
</script>
<script type="module">
  import * as m from '/examplemodule.js';
  karmaEs6Dependency.registerModule(m, 'example.module');
</script>
<script type="module">
  karmaEs6Dependency.postLoadingModule();
</script>
```

The example shows an ES6 module with `goog.module.declareNamespace`, which means it can be `goog.require`d. However that complicates things quite a bit. If a Closure file (e.g. a `goog.provide` or `goog.module` file, not an ES6 module) did `goog.require` that namespace, it must be loaded *after* the ES6 module it requires. If you load code with `script src="path"` this isn't an issue, just add `defer` (well, on IE11+ I think). But if you're putting raw text in a script tag (`<script>goog.loadModule/*...*/</script>`) then you can't `defer` that. You could make it a `type="module"`, which would work for `goog.module` files (since they'd just be calls to `goog.loadModule`), but it wouldn't work for `goog.provide` files, which can create globals. You could try having a `type="module"` that calls `goog.global` eval for those files.

So we might also have the following later the that html file:

```js
<script type="module">
  goog.loadModule(function(exports) {
    goog.module('example.goog.module');
    const exampleModule = goog.require('example.module');
    console.log('from goog.module: ' + exampleModule.filtered);
  });
</script>
<script type="module">
  // eval'd globally so as to ensure global values are global
  goog.globalEval(`
    goog.provide('example.provide');
    goog.require('example.module');
    var exampleGlobal = 0;
    console.log('from goog.provide: ' + goog.module.get('example.module').filtered);
  `);
</script>
```

A lot of this stuff doesn't have public documentation yet (the code is the current documentation). I wanted to vet this internally before publicly advertising it. Also our tools to parse files aren't up to date for ES6 modules yet (e.g. the stuff in closure/bin/build that generates deps.js files). I plan on writing a new node package to parse Closure files rather than continue editing python code. This new package will be available as a library as well, which may help you out. No need to parse things yourself.

Sorry for the wall of text. Documentation is coming, but only once things are vetted and we have the new node package to parse this stuff.

**Comment by nckh** [comment]
Thank you for the very thorough explanation. I'm going to need some time to review base.js's code and your suggestions.

Also for some reason I thought that "debug loader" and "code loader" were the same things. I suppose the debug loader is the mechanism that requires deps.js and make the browser load the raw source files in the right order, and the code loader is the process that transforms the code into $jscomp modules. Is this correct? 

Back to the Karma topic, I really want to avoid requiring any kind of bundling phase before running tests, as this would greatly hit our productivity. I may try to look around solutions to prevent Karma from loading the script files directly, and instead let base.js do it with a custom generated deps.js. No idea yet of the feasibility.

**Comment by nckh** [comment]
Thanks again for your help John.

I managed to implement the custom dependency factory as you recommended, but because I couldn't get the loading order working properly, and Karma not really allowing to write inline script tags, I ended up refactoring the preprocessor completely in order to delegate the actual dependency loading to the debug loader. https://github.com/nckh/karma-closure/tree/es6
